src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+# Introduction
+That document specifies alarm handling for thin-edge.
+
+# Data Semantic of Alarms
+Thin-edge treates alarms as **stateful signals**. Instead to usual **cyclic measured values** (as e.g. measurements of physical units like temperature, humidity, pressure, ...) a **stateful signal** is transfered only on state-change. If such a transfered state-change got lost, the cloud assumes the wrong state until a next state-change occurs. Especially in case of alarms that means a once raised alarm would completely unknown to the cloud.
+
+A lost alam-raise could be even more problematic if the raised alarm requires some cloud-site interaction (or manual interaction). Since the lost alarm is not visible on cloud no intercation will be started. And since not state-change happen again, the alarm will never appear at cloud.
+
+**Consequence:**<br/>
+All alarm state-changes need to be transfered reliable from **Application to thin-edge**, and from **thin-edge to cloud**.
+I.E. alarm state-changes must not got lost, or device software must be able to detect loss and react accordingly (e.g. retry to transfer).
+
+More details about loss off data-transfer and proper reaction see...  TODO: add data-flow-image here!
+
+# Information Set per Alarm
+The specified alarm attributes below are inspired by cumulocity data model for alarms, but shall be re-usable for other clouds where possible.
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.

```

Comment:
Data traffic would be less of a concern for alarms that are less frequently sent data, right? And it feels like the inherent request-response model of HTTP might make it easier to implement guaranteed alarm delivery to the cloud because of the responses.